### DISCUSSION TITLE
what happens when app server restarts constantly?

### DISCUSSION THREAD

**Discussion by karmingc** [petergoldstein/dalli#903] [Q&A] [answered]
For simplicity, let's assume there's 2 memcached servers and we constantly deploy new versions of our applications.

For reference, the implementation of the ring is: https://github.com/petergoldstein/dalli/blob/main/lib/dalli/ring.rb#L31

- Prior to app restarting, has one key (say "example") that maps to node 1.
- App restarts

is it possible that if the app builds the new cache key (say "example"), it maps to node 2?

**Comment by karmingc** [comment]
@petergoldstein i'm assuming everything will be lost in memory... would this be a possible scenario?

**Comment by petergoldstein** [comment, marked as answer]
@karmingc.  No, assuming you're using Dalli correctly, keys should map to the same node in different `Dalli::Client` instances in the same application server or in entirely different application servers.

The ring build is designed to be consistent.  That is, given the same set of memcached servers, in the same order, with the same weights, you should get the same continuum points for the ring.  You can see that [here](https://github.com/petergoldstein/dalli/blob/b330033febfe1c61c34e8936fbfa5d459ada20f5/lib/dalli/ring.rb#L128).  Similarly, the mapping of a key to a point on the continuum is designed to be deterministic as seen [here](https://github.com/petergoldstein/dalli/blob/b330033febfe1c61c34e8936fbfa5d459ada20f5/lib/dalli/ring.rb#L51).  The fact that the individual rings are in memory is irrelevant because they are rebuilt consistently.

The only situation where the mapping would change (given the same Dalli configuration for both application servers) is if the application servers don't see the same set of memcached servers.  That is, if one of the memcached servers appears down to one application server and up to another, then the mappings will differ.

And if you use different Dalli configurations for each application server, then you have no guarantees on consistency of mapping.

**Reply by karmingc** [reply]
@petergoldstein thank you for your elaborate answer! I was actually going to follow up with a question concerning kubernetes, but that answers as well. Cheers!
